build/root/Makefile

```diff
@@ -208,7 +208,7 @@ test-integration:
 	echo "$$TEST_IT_HELP_INFO"
 else
 test-integration:
-	hack/make-rules/test-integration.sh $(WHAT)
+	KUBE_TEST_ARGS='$(value KUBE_TEST_ARGS)' hack/make-rules/test-integration.sh $(WHAT)
```

Comment:
I don't really get what this is doing, which means the next time I need to touch this I will be ++lost.

As a quick level-set:

```
$ cat Makefile 
foo:
	echo $(VAR)

$  make
echo 

$ make VAR=foo
echo foo
foo

$ VAR=foo make
echo foo
foo

$ VAR=foo make VAR=bar
echo bar
bar
```

So I am missing what the specific problem being solved here is?

Reply:
The issue is the difference in how make and bash expand/escape variables. In particular, the dollar sign has different meanings.

In bash, when you want to prevent a dollar sign from having significant meaning you can escape it either by using a backslash or encompassing the entire value in single quotes which will prevent any variable expansion. In make, you have to double up the dollar signs.

`KUBE_TEST_ARGS` in this case can contain a regular expression which ALSO uses the dollar sign, but not as a variable marker. If you run your tests with something that has the dollar sign in them, I think you'll see what I mean. The dollar sign gets stripped. This is the issue that came up in https://github.com/kubernetes/community/pull/7724.

What this does is two fold:
- It explicitly orders the variable first before the bash script, so that the value is passed through the environment which avoids a loop through make's variable expansion layer
- It uses the `value` builtin of make to take the raw value of the variable without expanding it.

It took a bunch of fiddling to get this to work, but the result is, as long as you single quote the variable to avoid your own local shell mucking with it before it gets to make, it shouldn't matter the order you specify `KUBE_TEST_ARGS='-run ^TestPodUpdateActiveDeadlineSeconds$'`.. the dollar sign will be retained.

I also made changes to `hack/make-rules/test-integration.sh` to ensure that when we loop back through `make test` we are specifying `KUBE_TEST_ARGS` as an environment variable and unsetting `MAKEFLAGS` so that we don't go through another loop of variable expansion (what resulted in the inconsistent number of dollar signs in the community repo pull request above -- do I need two dollar signs if I prefix the variable, and four if I append it? -- ref https://github.com/kubernetes/community/pull/7724#discussion_r1494380146).

Reply:
I see.  So it's less about env vs CLI and more about preserving $ in the value.

Drop a comment to that effect and I will approve :)

Reply:
@thockin Done!